feat: add (optional) float16 support

manzt opened the pull request:
Fixes #127 

This PR implements optional support for `float16` datatypes by inspecting `globalThis.Float16Array`. Rather than adding https://github.com/petamoriken/float16 as a dependency, users must opt into this behavior by adding `Float16Array` as a global in their environment.

```javascript
import { openArray } from 'zarr';

const z = await openArray({ store: "http://example.com/f2data.zarr" });
await z.getRaw(null) // throws: '<f2' is not supported natively in zarr.js. In order to access this dataset you must make Float16Array available as a global. See https://github.com/gzuidhof/zarr.js/issues/127

import { Float16Array } from '@petamoriken/float16';
globalThis.Float16Array = Float16Array;
await z.getRaw(null)  // { data: Float16Array, ... }
```

manzt commented:
cc: @keller-mark

manzt commented:
From the typescript perspective, there is a more “correct” way of communicating this behavior, which I will add to this PR https://github.com/DefinitelyTyped/DefinitelyTyped/blob/fb66211010980641f01bcfe8ffbc0cc3d71989ca/types/ndarray/index.d.ts#L43

gzuidhof commented on src/nestedArray/types.ts:
```diff
@@ -1,6 +1,11 @@
 
+declare global {
+  // eslint-disable-next-line
+  var Float16Array: import('@petamoriken/float16').Float16ArrayConstructor;
```
I believe you can do `import type {Float16ArrayConstructor} from '@petamoriken/float16'` to only import the type and avoid something ending up in the bundle. I may be wrong but perhaps worth a try!

manzt replied:
The current babel config for jest does not support `type` imports like this, so this caused an issue. The inline type import `import('@petamoriken/float16').Float16ArrayConstructor;` is type-only and doesn't add anything to the bundle.

manzt replied:
This is no longer necessary, since we conditionally infer the type of `Float16Array` from globalThis.

gzuidhof reviewed (approved):
Really great :) I think it's the perfect way to support Float16Array without pulling in the library.

Perhaps this warrants a page in the docs? Even if it only contains the code snippet from OP it will already help I imagine.

gzuidhof commented on src/nestedArray/types.ts:
```diff
@@ -53,13 +62,21 @@ const DTYPE_TYPEDARRAY_MAPPING: { [A in DtypeString]: TypedArrayConstructor<Type
 
 export function getTypedArrayCtr(dtype: DtypeString) {
-  const ctr = DTYPE_TYPEDARRAY_MAPPING[dtype];
+  const ctr = DTYPE_TYPEDARRAY_MAPPING()[dtype];
```
I personally would've just kept it simple and require Float16Array to be defined prior to importing Zarr. Of course the performance differences are negligible, but it's the kind of thing where I tend to be strict rather than lenient

manzt replied:
I tried this, but I think due to how jest bundles/transpiles modules I am unable to set `Float16Array` globally in our test suite prior to importing zarr.js.

manzt replied:
Ok... actually I can do a dynamic import in the tests.

manzt commented:
Ok, I've changed the `TYPED_ARRAY_MAPPING` to inspect `globalThis` on import only, and also removed the global declaration from `zarr`. This way it is up to the end user to add the global declaration if they want the type safety.

![image](https://user-images.githubusercontent.com/24403730/188685045-726a1c0e-5050-439f-93e7-08ddb80001a2.png)

![image](https://user-images.githubusercontent.com/24403730/188685086-cb35b028-880f-4510-b1ca-5fa76199eb35.png)
